File: protobuf/server/golang/app/serverv1.go
package main

import (
	"context"
	pbv1 "grpc_tutorial/proto/v1"
	"log"
)

type serverGreeter struct {
	pbv1.UnimplementedGreeterServer
}

type serverDataStore struct {
	pbv1.UnimplementedDataStoreServer
}

var inMemoryMap = make(map[string]string)

// SayHello implements helloworld.GreeterServer
func (s *serverGreeter) SayHello(ctx context.Context, in *pbv1.HelloRequest) (*pbv1.HelloReply, error) {
	log.Printf("Received: %v", in.GetName())
	return &pbv1.HelloReply{Message: "Hello " + in.GetName()}, nil
}

func (s *serverDataStore) Save(ctx context.Context, in *pbv1.SaveRequest) (*pbv1.SaveReply, error) {
	log.Printf("Received: %v", in.GetKey())
	inMemoryMap[in.GetKey()] = in.GetValue()
	return &pbv1.SaveReply{Key: in.GetKey(), Value: in.GetValue()}, nil
}

func (s *serverDataStore) Load(ctx context.Context, in *pbv1.LoadRequest) (*pbv1.LoadReply, error) {
	log.Printf("Received: %v %v", in.GetKey(), inMemoryMap[in.GetKey()])
	if inMemoryMap[in.GetKey()] == "" {
		return &pbv1.LoadReply{Value: "Not found"}, nil
	}
	return &pbv1.LoadReply{Value: inMemoryMap[in.GetKey()]}, nil
}

File: protobuf/server/python/app/servers/serverv1.py
from meysam_utils import get_logger

logger = get_logger(__name__)


class Greeter(greetings_pb2_grpc.GreeterServicer):
    def SayHello(self, request, context):
        logger.info("Greeter server received: " + request.name)
        try:
            from app.tasks import app as celery_app

            celery_app.send_task("greet", args=[request.name])
        except Exception as e:
            logger.error(e)
        return greetings_pb2.HelloReply(message="Hello %s" % request.name)


class DataStore(data_store_pb2_grpc.DataStoreServicer):
    data_store = {}

    def Save(self, request, context):
        logger.info("DataStore server received: " + request.key)
        self.data_store[request.key] = request.value
        return data_store_pb2.SaveReply(key=request.key, value=request.value)

    def Load(self, request, context):
        logger.info("DataStore server received: " + request.key)
        value = self.data_store.get(request.key, "Not found")
        return data_store_pb2.LoadReply(value=value)


services = [
    {"service": Greeter, "handler": greetings_pb2_grpc.add_GreeterServicer_to_server},
    {
        "service": DataStore,
        "handler": data_store_pb2_grpc.add_DataStoreServicer_to_server,
    },
]
